subsys/fs/zms/zms_priv.h

```diff
@@ -0,0 +1,66 @@
+/*  ZMS: Zephyr Memory Storage
+ *
+ * Copyright (c) 2024 BayLibre SAS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef __ZMS_PRIV_H_
+#define __ZMS_PRIV_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*
+ * MASKS AND SHIFT FOR ADDRESSES
+ * an address in zms is an uint64_t where:
+ *   high 4 bytes represent the sector number
+ *   low 4 bytes represent the offset in a sector
+ */
+#define ADDR_SECT_MASK  GENMASK64(63, 32)
+#define ADDR_SECT_SHIFT 32
+#define ADDR_OFFS_MASK  GENMASK64(31, 0)
+#define SECTOR_NUM(x)   FIELD_GET(ADDR_SECT_MASK, x)
+
+#if defined(CONFIG_ZMS_CUSTOM_BLOCK_SIZE)
+#define ZMS_BLOCK_SIZE CONFIG_ZMS_MAX_BLOCK_SIZE
+#else
+#define ZMS_BLOCK_SIZE 32
+#endif
+
+#define ZMS_LOOKUP_CACHE_NO_ADDR GENMASK64(63, 0)
+#define ZMS_HEAD_ID              GENMASK(31, 0)
+
+#define ZMS_VERSION_MASK    GENMASK(7, 0)
+#define ZMS_GET_VERSION(x)  FIELD_GET(ZMS_VERSION_MASK, x)
+#define ZMS_DEFAULT_VERSION 1
+
+/* Allocation Table Entry */
+struct zms_ate {
+	uint32_t id;  /* data id */
+	uint16_t len; /* data len within sector */
+	union {
+		struct {
+			uint32_t offset; /* data offset within sector */
+			union {
+				uint32_t data_crc; /* crc for data */
+				uint32_t metadata; /* Used to store metadata information
+						    * such as storage version.
+						    */
+			};
+		};
+		uint8_t data[8]; /* used to store small size data */
+	};
+	uint8_t cycle_cnt; /* cycle counter for non erasable devices */
```

Comment:
I know there it not enough place in ate for that, but should we maybe consider somehow storing total, not rounded to [0,255] counter in some special ate/empty sector ate?
Assuming that there will be flawless operation and this information does not get lost, we can then pick, from device, information on how much life is left in it.

Reply:
Actually we can store it in the close ATE metadata 32 bits.
But we need to make sure to keep the cycle count for devices that need an erase because the cycle count for a sector will be overwritten